File: android/src/main/java/com/sendbird/calls/reactnative/RNSBGroupCallVideoView.kt
package com.sendbird.calls.reactnative

import android.content.Context
import com.sendbird.calls.ParticipantState
import com.sendbird.calls.Room
import com.sendbird.calls.SendBirdCall
import com.sendbird.calls.reactnative.view.BaseVideoView

class RNSBGroupCallVideoView(context: Context) : BaseVideoView(context) {
    private var mRoomId: String? = null
    private var mParticipantId: String? = null
    private var mParticipantState = ParticipantState.ENTERED

    private fun getRoom(id: String?): Room? {
        val roomId = id ?: mRoomId ?: return null
        return SendBirdCall.getCachedRoomById(roomId)
    }

    fun updateView() {
        val participant = this.getRoom(mRoomId)?.participants?.find { participant -> participant.participantId == mParticipantId }
        participant?.videoView = mSurface
    }

    fun setParticipantId(participantId: String) {
        this.mParticipantId = participantId
        this.updateView()
    }

    fun setRoomId(roomId: String) {
        this.mRoomId = roomId
        this.updateView()
    }

    fun setParticipantState(state: ParticipantState) {
        this.mParticipantState = state
        this.updateView()
    }
}

File: ios/RNSBGroupCallVideoView.swift
//
//  RNSBGroupCallVideoView.swift
//  RNSendbirdCalls
//
//  Created by James Kim on 2022/07/12.
//  Copyright © 2022 Facebook. All rights reserved.
//

import Foundation
import React
import SendBirdCalls

@objc(RNSBGroupCallVideoView)
class RNSBGroupCallVideoView: BaseVideoView {
    @objc var roomId: String? = nil
    @objc var participantId: String? = nil
    @objc var state: String = SendBirdCalls.Participant.State.entered.asString
    
    private func getRoom(id: String?) -> Room? {
        guard let roomId = roomId else { return nil }
        
        if let room = SendBirdCall.getCachedRoom(by: roomId) {
            return room
        }
                
        return nil
    }
    
    
    override func didSetProps(_ changedProps: [String]!) {
        super.didSetProps(changedProps)
        
        if changedProps.contains("roomId") {
            updateView()
        }
        
        if changedProps.contains("participantId") {
            updateView()
        }
        
        if changedProps.contains("state") {
            if SendBirdCalls.Participant.State.init(fromString: self.state.uppercased()) != nil {
                updateView()
            }
        }
    }
    
    private func updateView() {
        guard let room = getRoom(id: self.roomId) else { return }
        guard let participant = room.participants.first(where: { $0.participantId == self.participantId }) else { return }
        participant.videoView = self.surface
    }

}
